### PR TITLE
Implement ext_proc header allow list feature

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
With the header allow list config, only allowed headers in either request or response headers or trailers are forwarded to the ext_proc server.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/extensions/filters/http/ext_proc/mutation_utils_test.cc
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_proc/mutation_utils_test.cc:248]: Just tested EXACT case to verify the headersToProto() utility function is honoring the new header_matchers parameter, and working as expected.  The complete matching function test is in this file:
https://github.com/envoyproxy/envoy/blob/c2ae2211196a48b12d2e36d00c6c2889ae2f434a/test/common/common/matchers_test.cc#L315

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (yanjunxiang-google) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:2026]: This test verifies  only allowed headers of request or response headers | trailers are sent to the ext_proc server.

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto
Comment (stevenzzzz) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:192]: Does a disallow list make sense? 
often times we are doing filtering based on blacklist: remove some internal headers (e.g. x-envoy-xxx)  and sent everything else to the external service.

Comment (yanjunxiang-google) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:192]: Should we enforce allow list or disallow list is discussed in this PR:
https://github.com/envoyproxy/envoy/pull/27333

**Comment by yanavlasov** [comment]
Format is unhappy about trailing space: https://dev.azure.com/cncf/envoy/_build/results?buildId=137902&view=logs&j=c7de00aa-b0fb-52ba-707b-da6afd918eda&t=2ec0901d-43fb-521a-88ec-473706b56eab&l=31

/wait

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h]: done

File: source/extensions/filters/http/ext_proc/mutation_utils.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/mutation_utils.cc:38]: For observability purpose, we probably should keep track of the headers that are disallowed. So. I think logging them at least

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/ext_proc.h:89]: nit: `initHeadMatchers` ---- > `initHeaderMatchers`

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (tyxia) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:2050]: if `handleUpstreamRequestWithTrailer` is not called, it is expected that no trailer is there even without allowList rule, right?
Probably a better test will be: encode the other trailer besides `x-test-trailers` and verify that is not there.

**Review by yanavlasov — APPROVED** [review]

**Comment by yanavlasov** [comment]
LGTM, waiting for API approval.

File: source/extensions/filters/http/ext_proc/mutation_utils.cc
Comment (tyxia) [inline source/extensions/filters/http/ext_proc/mutation_utils.cc:27]: For the case that multiple ext_proc filters in the chain with same forwarding rule, we probably can cache and share the match result.

But this is out of scope of this PR and it's only safe when forwarding rule is same (i.e., with indication from config). We can think about/discuss this later.

**Review by tyxia — COMMENTED** [review]
Thank you for working on this!

Mostly LGTM, just a few comments.

**Review by htuch — COMMENTED** [review]
/lgtm api

File: source/extensions/filters/http/ext_proc/mutation_utils.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:27]: Yeah, we can discuss that implementation optimization later.

File: source/extensions/filters/http/ext_proc/ext_proc.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/ext_proc.h:89]: done

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:2024]: :status is to have this response header be able to sent to the ext_proc server.

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (yanjunxiang-google) [reply test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:2050]: codec_client->sendTrailers("x-trailer-foo) will send downstream request trailers to ext_proc filter, and be forwarded to ext_proc server. However, since "x-trailer-foo" is not in the allow header list, so the trailer forwarded to ext_proc server doesn't contain any header.  This is tested and verified the allow header feature for the request trailer.

With handleUpstreamRequestWithTrailer(), the upstream will send back and response with trailer which contains a fixed header: x-test-trailers. Since it is in allow list, it will be forwarded to the ext_proc server. And ext_proc server verified this. This tested the allow header feature for the response trailer.

Combine this two, it tested the scenario you mentioned. Also, with the header allow list for request / response headers, I feel the test is comprehensive. Otherwise, we will have to modify handleUpstreamRequestWithTrailer() a little bit to add more headers there, but is not necessary.

File: source/extensions/filters/http/ext_proc/mutation_utils.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:35]: for 1,  std::vector is used in similar filter implementations, so let's keep them consistent.
for 2, done

File: source/extensions/filters/http/ext_proc/mutation_utils.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:38]: Since we are doing allow list, thus there could be significant amount of headers being disallowed for each request. Logging here could be very noisy.

File: source/extensions/filters/http/ext_proc/mutation_utils.h
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.h:21]: done

File: source/extensions/filters/http/ext_proc/mutation_utils.cc
Comment (yanjunxiang-google) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:27]: > Thank you for working on this!
> 
> Mostly LGTM, just a few comments.

Thanks for the comments!

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc
Comment (tyxia) [reply test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:2050]: ah, I missed single line `sendTrailers("x-trailer-foo)` above

File: source/extensions/filters/http/ext_proc/mutation_utils.cc
Comment (tyxia) [reply source/extensions/filters/http/ext_proc/mutation_utils.cc:38]: On second thought, the allowList is specified by the end user. So it should be fine from end user observability perspective.

**Review by tyxia — APPROVED** [review]
Thanks!

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**
